utils/blockhash.mjs

```diff
@@ -1,64 +1,97 @@
-          5000
-        )
-      );
-      // Get the latest blockhash from the RPC node and update the global
-      // blockhash object with the new value and timestamp. If the RPC node
-      // fails to respond within 5 seconds, the promise will reject and the
-      // script will log an error.
-      const latestBlockhash = await Promise.race([
-        rpc.getLatestBlockhash().send(),
-        timeoutPromise,
-      ]);
-
-      gBlockhash.value = latestBlockhash.value.blockhash;
-      gBlockhash.lastValidBlockHeight = latestBlockhash.value.lastValidBlockHeight;
+const RECENT_BLOCKHASHES_ADDRESS = address(
+  "SysvarRecentB1ockHashes11111111111111111111"
+);
 
-      gBlockhash.updated_at = Date.now();
-      attempts = 0;
-    } catch (error) {
-      gBlockhash.value = null;
-      gBlockhash.updated_at = 0;
+async function getDifferenceBetweenSlotHeightAndBlockHeight() {
+  const { absoluteSlot, blockHeight } = await rpc
+    .getEpochInfo()
+    .send({ abortSignal: AbortSignal.any([]) });
+  return absoluteSlot - blockHeight;
+}
 
-      ++attempts;
+function getLatestBlockhashFromNotification(
+  {
+    context: { slot },
+    value: {
+      data: {
+        parsed: {
+          info: [{ blockhash }],
+        },
+      },
+    },
+  },
+  differenceBetweenSlotHeightAndBlockHeight
+) {
+  return {
+    blockhash,
+    lastValidBlockHeight: (slot - differenceBetweenSlotHeightAndBlockHeight) + 150n,
+  };
+}
 
-      if (error.message.includes("new blockhash")) {
-        console.log(
-          `${new Date().toISOString()} ERROR: Unable to obtain a new blockhash`
+let differenceBetweenSlotHeightAndBlockHeight;
+let latestBlockhash;
+let recentBlockhashesNotifications;
+async function* blockhashes() {
+  if (latestBlockhash !== undefined) {
+    yield latestBlockhash;
+  }
+  if (recentBlockhashesNotifications === undefined) {
+    [
+      differenceBetweenSlotHeightAndBlockHeight,
+      recentBlockhashesNotifications,
+    ] = await safeRace([
+      Promise.all([
+        getDifferenceBetweenSlotHeightAndBlockHeight(),
+        rpcSubscriptions
+          .accountNotifications(RECENT_BLOCKHASHES_ADDRESS, {
+            encoding: "jsonParsed",
+          })
+          .subscribe({ abortSignal: AbortSignal.any([]) }),
+      ]),
+      // If the RPC node fails to respond within 5 seconds, throw an error.
+      timeout(5000),
+    ]);
+    // Iterate over the notificatons forever, constantly updating the `latestBlockhash` cache.
+    (async () => {
+      for await (const notification of recentBlockhashesNotifications) {
+        latestBlockhash = getLatestBlockhashFromNotification(
+          notification,
+          differenceBetweenSlotHeightAndBlockHeight
         );
+      }
+    })();
+  }
+  for await (const notification of recentBlockhashesNotifications) {
+    latestBlockhash = getLatestBlockhashFromNotification(
+      notification,
+      differenceBetweenSlotHeightAndBlockHeight
+    );
+    yield latestBlockhash;
+  }
```

Comment:
You essentially want a read-through cache that has the freshest blockhash at all times. Watching the recent blockhashes sysvar is a fun way to do this, so long as you don't mind the sort of gross computation of `lastValidBlockHeight`.